### PR TITLE
Periodic warning for 1-node cluster w/ seed hosts

### LINKED ISSUE #85222: More helpful logs if a single-node cluster has nontrivial discovery config
Users sometimes accidentally auto-bootstrap their nodes into single-node clusters and then try and join them all together later by setting `discovery.seed_hosts` or similar. Elasticsearch doesn't care if discovery points to a few extra nodes so today it reports no problems, but it's frustrating to the user that it doesn't do what they expect it to do and also doesn't indicate why.

It's almost certainly incorrect to have a nontrivial discovery config in a single-node cluster: `discovery.seed_hosts` and `discovery.seed_providers` should both be unset or empty[^1], so this is a strong indicator that the user is trying to form a multi-node cluster incorrectly. We can therefore be more helpful to these users by emitting periodic warnings in a single-node cluster with nontrivial discovery config:

> `This node is a fully-formed single-node cluster with cluster UUID [{clusterUUID}], but [{setting}] is set to [{value}] as if to discover other nodes and form a multi-node cluster. Fully-formed clusters do not attempt to discover other nodes, and nodes with different cluster UUIDs cannot belong to the same cluster. The cluster UUID persists across restarts and can only be changed by deleting the contents of the node's data path(s). Remove the discovery configuration to suppress this message.`

[^1]: it's also (kind of) acceptable if `discovery.seed_hosts` contains just the node's own publish address, or we're using just the `file` provider and `unicast_hosts.txt` contains just the node's own publish address.

### PR DESCRIPTION
**Pull request by kingherc**
For fully-formed single-node clusters, emit a periodic warning if
seed hosts have been configured.

Closes #85222

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:120]: Settings must be registered in `ClusterSettings` otherwise they won't be accepted in the config file:

```diff
diff --git a/server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java b/server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java
index b225f6fb4ca..89d16c01df4 100644
--- a/server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java
+++ b/server/src/main/java/org/elasticsearch/common/settings/ClusterSettings.java
@@ -492,6 +492,7 @@ public final class ClusterSettings extends AbstractScopedSettings {
         ElectionSchedulerFactory.ELECTION_DURATION_SETTING,
         Coordinator.PUBLISH_TIMEOUT_SETTING,
         Coordinator.PUBLISH_INFO_TIMEOUT_SETTING,
+        Coordinator.SINGLE_NODE_CLUSTER_SEED_HOSTS_CHECK_INTERVAL_SETTING,
         JoinValidationService.JOIN_VALIDATION_CACHE_TIMEOUT_SETTING,
         FollowersChecker.FOLLOWER_CHECK_TIMEOUT_SETTING,
         FollowersChecker.FOLLOWER_CHECK_INTERVAL_SETTING,
```

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left some suggestions/comments but nothing major.

**Comment by DaveCTurner** [comment]
Also I've added the `cloud-deploy` label, it would be good to confirm (manually) that this doesn't trigger any warnings on single-node Cloud clusters.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:840]: Indeed it cannot happen, I will remove this `if`.

File: server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java:853]: Nice, I'll do this. Could you tell me the usefulness of the `Set.copyOf` here?

**Comment by kingherc** [comment]
Hi @DaveCTurner , I've manually checked that the warning doesn't appear on a local distribution I built. How can I manually test it on a single-node cloud cluster? Does the cloud-deploy label somehow automatically spin a cloud cluster so I can check the logs?

**Comment by kingherc** [comment]
Hi @DaveCTurner , also pushed suggested changes, please check once more.

**Comment by DaveCTurner** [comment]
> How can I manually test it on a single-node cloud cluster?

I shared a link to the internal docs in another channel.

**Comment by kingherc** [comment]
@DaveCTurner please check once more and review.

**Comment by kingherc** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

**Comment by kingherc** [comment]
@elasticmachine run elasticsearch-ci/packaging-tests-unix-sample

**Comment by kingherc** [comment]
Hi @DaveCTurner , PR is ready for your review.

**Comment by DaveCTurner** [comment]
:+1: sorry I'm unlikely to get to this today now, but it's on my list.

File: docs/reference/modules/discovery/discovery-settings.asciidoc
Comment (DaveCTurner) [inline docs/reference/modules/discovery/discovery-settings.asciidoc:208]: Reflecting on the user-facing view of this feature, I think something more generic would be a bit better. For instance:

```suggestion
`cluster.discovery_configuration_check.interval `::
(<<static-cluster-setting,Static>>)
Sets the interval of some checks that will log warnings about an
incorrect discovery configuration. The default value is `30s`.
```

Not that we necessarily intend to add any more periodic checks in future, but if we did we could use the same schedule.

**Review by DaveCTurner — COMMENTED** [review]
Changes all look good except for one tiny comment about naming/docs. However I think we should consider this blocked on https://github.com/elastic/cloud-on-k8s/issues/5834. Not sure if you should (or want to) fix that too.

**Comment by kingherc** [comment]
@DaveCTurner I was eager to merge this but I understand we want to make sure cloud sets the settings correctly :) 
By the way, when I tested it, I did see that seed_hosts was set to []. So the https://github.com/elastic/cloud-on-k8s/issues/5834 is to make sure it truly happens and it's not by chance?

**Comment by kingherc** [comment]
@DaveCTurner  thinking more about this: The current condition for the warning is if the seed_hosts is set, and it is non-empty. I think the cloud uses either the file method (without the seed_hosts at all) or might set it to an empty list. So I think the current code should be OK for the cloud. Or do they populate seed_hosts for some reason for single-node clusters? (I remind we don't look now at all the last discovered hosts)

**Comment by DaveCTurner** [comment]
I think this works fine as-is in Cloud (i.e. ESS) but not in ECK.

**Comment by kingherc** [comment]
@DaveCTurner , ah I see! I think ECK does not set seed_hosts at all, so they should not get the warning (as we do not look now for last discovered hosts -- which would include the unicast_hosts.txt file). I can test ECK with a single-node to confirm. Will try to see if they have instructions on testing.

**Review by DaveCTurner — APPROVED** [review]
Aha, right, thanks for the correction. I lost track of this in all the iterations. You're right, this should be fine for both ECK and ESS.

**Comment by kingherc** [comment]
Thanks @DaveCTurner ! By the way, I believe the issue https://github.com/elastic/cloud-on-k8s/issues/5834 might not be needed anymore with our new approach. Should we propose to close it?

**Comment by DaveCTurner** [comment]
No I think the ECK issue still has value even now. If a node cannot find the master then it will (eventually) emit log messages about all the addresses it's trying, which will include the default seed addresses if this setting isn't set, and that just causes confusing noise.

**Comment by kingherc** [comment]
Ah, that's another message that gets emitted while trying to find masters. OK, thanks for the explanation! Will merge this when it passes the tests.

**Pull request merged**
